### ISSUE TITLE
[css-scoping] Consider aligning ::slotted() for fallback content with implementations

### ISSUE THREAD

**Issue by lilles** [w3c/csswg-drafts#5482]
The scoping spec states that:

"The ::slotted() pseudo-element represents the elements assigned, after flattening"

According to the flattening algorithm in the HTML spec, that also includes the fallback children of a slot if that slot has no assigned nodes. This behavior was agreed on in [WebComponents issue 631](https://github.com/w3c/webcomponents/issues/631) and issues were reported for Blink and WebKit, but have not been fixed.

The current implementations walk the assignedSlot chain and apply ::slotted rules from each slot's scope, assignedSlot being null for fallback elements.

This is an example of how this works (in all browsers per current spec) with nested slot assignment:

```
<!doctype html>
<div id="host"><span class="a">Orange with blue border</span></div>
<script>
  const root = host.attachShadow({mode:"open"});
  root.innerHTML = `
    <style>
      ::slotted(.a) { border: 2px solid blue }
    </style>
    <div id="host">
      <slot></slot>
    </div>
  `;
  const inner_root = root.querySelector("#host").attachShadow({mode:"open"});
  inner_root.innerHTML = `
    <style>
      ::slotted(.a) { color: orange; }
    </style>
    <slot></slot>
  `;
</script>
```

Per current spec, moving the span into fallback content for the slot should also result in orange text and blue border, but this is unstyled in all browsers):

```
<!doctype html>
<div id="host"></div>
<script>
  const root = host.attachShadow({mode:"open"});
  root.innerHTML = `
    <style>
      ::slotted(.a) { border: 2px solid blue }
    </style>
    <div id="host">
      <slot><span class="a">Orange with blue border</span></slot>
    </div>
  `;
  const inner_root = root.querySelector("#host").attachShadow({mode:"open"});
  inner_root.innerHTML = `
    <style>
      ::slotted(.a) { color: orange; }
    </style>
    <slot></slot>
  `;
</script>
```

If the author wants to style a certain type of element the same whether it is assigned, or fallback, it can be accomplished with the current implementations adding an extra selector like this:

```::slotted(.a), slot > .a```

However, that only works in the scope where the fallback is. For the example above, you can do this for the tree with the border rule, but not when that slot is assigned to the slot of the inner root.

I have done an implementation which matches the current spec in Blink, and it is pretty straightforward. However, compat is a big concern here. The impact of changing the implementation would be hard to use-count/quantify. I've been in contact with engineers at Google working on Web Components and Polymer who thinks that changing this is too risky compat-wise and would prefer a spec change.

This surfaced in a [bug report](https://crbug.com/1119248) from Salesforce which found inconsistencies between implementations, but it turns out that is about ::slotted(slot) matching nested slots in WebKit, while Blink and Firefox do not match those:

```
<!doctype html>
<div id="host"><span class="a" slot="inner">Orange with blue border</span></div>
<script>
  const root = host.attachShadow({mode:"open"});
  root.innerHTML = `
    <style>
      ::slotted(.a) { border: 2px solid blue }
    </style>
    <div id="host">
      <slot>
        <slot name="inner"></slot>
      </slot>
    </div>
  `;
  const inner_root = root.querySelector("#host").attachShadow({mode:"open"});
  inner_root.innerHTML = `
    <style>
      ::slotted(slot) { color: orange; }
    </style>
    <slot></slot>
  `;
</script>
``` 

@emilio @rniwa @gregwhitworth

**Comment by gregwhitworth** [comment]
The impact of changing the implementation would be hard to use-count/quantify.

Can we start at least with utilization of `::slotted()` and it being truly evoked via usecounter? I agree that mapping it default content fallback will be a bit more work.

Beyond that I think we shouldn't prolong checking on the compat because the compat will simply continue to grow. If we do discover major compat implications (a lot of unique utilization from different sources, not solely traffic) we can do a measured release with a lot of communication around it. If it seems the issue has proliferated beyond where outreach may be of help (eg: it isn't 5-10 major web teams/libs) we can then discuss how we want to handle compat.

**Comment by lilles** [comment]
Use counter for ::slotted():

https://www.chromestatus.com/metrics/feature/timeline/popularity/1118

**Comment by rniwa** [comment]
FWIW, WebKit bug about this is https://bugs.webkit.org/show_bug.cgi?id=169948 but we have no immediate plan to take actions on this.

**Comment by gregwhitworth** [comment]
Thanks @lilles much appreciated. So comparing `::slotted()` with something like `focus-visible` which is relatively new:

`focus-visible` 2.18%
<img width="645" alt="Screen Shot 2020-09-01 at 5 05 56 PM" src="https://user-images.githubusercontent.com/865244/91917697-84b4ed80-ec75-11ea-907e-3aa4fda5cc84.png">

`::slotted()` .19%
<img width="646" alt="Screen Shot 2020-09-01 at 5 07 28 PM" src="https://user-images.githubusercontent.com/865244/91917749-ae6e1480-ec75-11ea-964d-301b6d81d1f7.png">

While `::slotted()` is used, as @lilles  noted, this doesn't account for if it is default content or not and my hunch is that most don't target default content. I think, if feasible, is for Chrome (since they have a patch ready) to ship to canary and dev and see if bug reports come in from the change. If we don't, then potential compat concerns will continue to rise.

I'd love to see us at the very least try and determine if the compat risk is large, which currently being well under 1%, while not inconsequential I do think we're at a point this is a worthwhile change. If we don't then we'll need to add some type of switch to allow for default content to achieve the behavior as currently specified.

**Comment by tabatkins** [comment]
I'll note that having `::slotted()` target fallback contents was not an explicit goal when I was writing the spec text; I just wrote the simplest thing that did the job using the tree constructs I had available. I don't think I have an opinion on which way we go with this; the argument that you'll probably want to style fallback differently from slotted content sounds reasonable to me, tho (and thus having `::slotted()` *not* match would be better).

Looking into the details a bit more, it looks like the spec might actually be wrong, and not match any implementations anyway! In particular, I'm using ["find slottables"](https://dom.spec.whatwg.org/#find-slotables) to produce the flattened tree (which is correct), and that means the inner shadow root in Rune's original example will have a `<slot>` (from the outer shadow root) as its direct child in the flat tree; the `span.a` element will be a grandchild instead, and is not "assigned, after flattening, to the slot that is ::slotted’s originating element" (instead it's just carried along by its parent, the outer shadow's `<slot>` that *is* "assigned, after flattening"), and thus won't be selected by `::slotted(.a)` in the inner shadow.

That's probably related to:

> This surfaced in a bug report from Salesforce which found inconsistencies between implementations, but it turns out that is about ::slotted(slot) matching nested slots in WebKit, while Blink and Firefox do not match those:

Because nested slots *should*, per spec, be matched by `::slotted(slot)`.  But apparently WebKit *also* colors Rune's first example orange, indicating they're *also* matching the `.a` grandchild of the inner slot???

Basically I'm very confused about what browsers are doing and am open to clarifying the spec any which way.

**Comment by tabatkins** [comment]
Looking at this again, yeah, I'm pretty sure I need to, somehow, invoke [find flattened slotables](https://dom.spec.whatwg.org/#find-flattened-slotables) for the purpose of gathering the elements that'll be matched by the `::slotted()`'s selector argument.

That'll mean that `slot` elements never match `::slotted(slot)`; this won't match Safari, but it'll match other parts of their behavior, and match Chrome and Firefox, and I think match the intended mental model better.
